File: May-LeetCoding-Challenge/31-Edit-Distance/Edit-Distance.cs
public class Solution {
    public int MinDistance(string word1, string word2) {
            int m = word1.Length;
            int n = word2.Length;
            int[,] dp = new int[2, n + 1];

            for (int i = 0; i <= n; i++)
                dp[0, i] = i;

            for (int i = 1; i <= m; i++)
            {
                for (int j = 0; j <= n; j++)
                {
                    if (j == 0)
                        dp[i % 2, j] = i;

                    else if (word1[i - 1] == word2[j - 1])
                        dp[i % 2, j] = dp[(i - 1) % 2, j - 1];

                    else
                        dp[i % 2, j] = 1 + Math.Min(Math.Min(dp[i % 2, j - 1], // Insert 
                                           dp[(i - 1) % 2, j]), // Remove 
                                           dp[(i - 1) % 2, j - 1]); // Replace 
                }
            }

            return dp[m % 2, n];
    }
}

File: May-LeetCoding-Challenge/31-Edit-Distance/Edit-Distance.java
/**
 * Dynamic programming, operation of previous string plus one if the last character is not the same
 * XXXb -> YYYc => operation(XXX to YYY) + 1
 * XXXm -> XXXm => operation(XXX to YYY)
 */
class Solution {
    public int minDistance(String word1, String word2) {
        int m = word1.length(), n = word2.length();
        int dp[][] = new int[m+1][n+1];

        /**
         * Construct a dp table
         * __|''|a|b|c|
         * ''| 0|1|2|3|...
         * ''| 0
         * d | 1
         * e | 2
         * f | 3
         */
        for (int i = 0; i <= m; i++) {
            dp[i][0] = i;
        }

        for (int j = 0; j <= n; j++) {
            dp[0][j] = j;
        }

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                char c1 = word1.charAt(i);
                char c2 = word2.charAt(j);

                // If the last character is the same, we do nothing compare to previous string
                if (c1 == c2) {
                    dp[i + 1][j + 1] = dp[i][j];
                } else {
                    // we chose the minimum operation of three operation: insert, delete, replace
                    dp[i + 1][j + 1] = Math.min(dp[i][j], Math.min(dp[i+1][j], dp[i][j+1])) + 1;
                }
            }
        }

        return dp[m][n];
    }
}